[nknguyenhc] iP

nknguyenhc opened the pull request:
# Quack-NKN bot

> "Whenever you win, you remember my name." - Nguyen :grin:

This Duke bot is so boring! Only a greeting line! I am suggesting major improvements to the bot. Here is a non-exhaustive list of improvements:
* Add a to-do, deadline or event to the current list of task
* View the current list of task, with filters applied if indicated
* Delete a task from the list
* Load data from and save data to disk

## Setup
Here are the steps to contribute to my fork:
1. Clone my repo `git clone https://github.com/nknguyenhc/ip.git`.
1. Open Intellij and do Java SDK setup.
1. In `build.gradle` file, click the play button besides the `run` command. Alternatively, double click `Ctrl` key and run `gradlew run`.
Congratulations! You have run the bot locally. Now you can start contributing.

## Releases
1. Click on my release tab and download the binary file.
1. In the terminal, redirect to the folder that contains the binary file.
1. Run `java -jar duke.jar`

## Notes
Last but not least, **ensure that you have done the following**:
- [X] Download my release and try it out.
- [ ] Visit my [**personal website**](https://nknguyenhc.github.io/).
- [ ] Try out my [**Orbital app**](https://matchminer-d5ebcada4488.herokuapp.com/about).

## Java
My project uses Java, which is a heavily object-oriented language. If you do not know Java, here is a block of code for you to get started:
```java
public class Quack {
    public static void main(String[] args) {
        System.out.println("quack");
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

nknguyenhc committed:
Level 0

Rename, greet and exit.

nknguyenhc committed:
Level 1

Echo

nknguyenhc committed:
Level 2

Add, list

nknguyenhc committed:
Level 3: Mark as Done

nknguyenhc committed:
Level 4

Todo, Event, Deadline

nknguyenhc committed:
Automated Tests

Tested mark as done and task types

nknguyenhc committed:
Error Handling

More extensive error handling on adding task. Test cases written.
If-else statements in interact changed to switch statement

nknguyenhc committed:
Documentation

nknguyenhc committed:
Level 6 Delete task

nknguyenhc committed:
Exclude test results

nknguyenhc committed:
Coding conventions

Changed static constant name, indicated constants as final

nknguyenhc committed:
Save data to hard disk

nknguyenhc committed:
Merge branch 'branch-Level-7'

nknguyenhc committed:
Parse & display datetime

nknguyenhc committed:
Stretch dates

List based on dates and task type

nknguyenhc committed:
Merge branch 'branch-Level-8'

nknguyenhc committed:
TaskList class

nknguyenhc committed:
Reorganised DateParseException

nknguyenhc committed:
Storage Class

nknguyenhc committed:
Parse and UI

Bug fixes as well

nknguyenhc committed:
Merge branch 'A-MoreOOP'

nknguyenhc committed:
Packaging

nknguyenhc committed:
Update Parser.java

nknguyenhc committed:
Merge branch 'add-gradle-support'

nknguyenhc committed:
Change package name in gradle

nknguyenhc committed:
Unit Tests

Added equals method to all classes implementing Command, for testing purpose

nknguyenhc committed:
Coding standards

nknguyenhc committed:
Docs

nknguyenhc committed:
Add task count notif

nknguyenhc committed:
Find task

nknguyenhc committed:
Merge branch 'branch-A-CodingStandard'

nknguyenhc committed:
Merge branch 'branch-A-JavaDoc'

nknguyenhc committed:
Merge branch 'branch-Level-9'

nknguyenhc committed:
Check style

marquestye commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,226 @@
+     * Date format: either "today", "tmr", "tomorrow", or DD/MM/YYYY
+     * Time format: either "{HH:MM}", "{HH}am", "{HH}pm", "{HH:MM}am" or "{HH:MM}pm"
+     */
+    private void interact() {
```
Great effort on your very descriptive JavaDoc, even for private methods!

marquestye commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,226 @@
+    /**
+     * Instantiate the bot, and starts the interaction immediately.
+     */
+    public Duke() {
```
Remember to start the first sentence with "Instantiates" rather than "Instantiate", as mentioned [here](https://se-education.org/guides/conventions/java/intermediate.html#comments).

marquestye commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,226 @@
+     * If there is IO error, handle the exception gracefully,
+     * and continue with an empty task list.
+     */
+    private void readFromDisk() throws Storage.FileCorruptedException {
```
Perhaps the first and second line can be swapped, as the second line is a better summary of what the method is doing. Swapping will also follow the coding standard on the first word's form.

marquestye commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,226 @@
+     * @param date date to display the deadlines before and events happening on,
+     *             null if to not exclude any task by date
+     */
+    public void showList(boolean isExcludingDone, LocalDate date) {
```
The coding standard also suggests the following:
- Empty line between description and parameter section.
- Punctuation behind each parameter description.

The words "whether" and "date" (following the parameters) should be capitalised.

marquestye commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,226 @@
+     * Adds a task to
+     * @param task
+     */
+    public void addTaskToList(Task task) {
```
Oops, this one seems to be incomplete.

marquestye commented on src/main/java/duke/parse/DateTimeManager.java:
```diff
@@ -0,0 +1,204 @@
+            }
+            if (possibleAmPm.equals("am")) {
+                isAm = true;
+            }
```
Instead of having nested if statements, could these be simplified to:
```
isAm = possibleAmPm.equals("am");
isPm = possibleAmPm.equals("pm");
```

nknguyenhc committed:
Refactor & show list in GUI

nknguyenhc committed:
Adjust appearance & layout

nknguyenhc committed:
Complete GUI

nknguyenhc committed:
Adjust appearance

Includes:
- Bot to monospaced font
- Combine messages
- Constant scrolling
- Style check

marquestye commented on src/main/java/duke/parse/Parser.java:
```diff
@@ -0,0 +1,223 @@
+    public static Command parse(String input) throws ParseError {
+        String[] commandArgs = input.split(" ", 2);
+        int index;
+        switch (commandArgs[0]) {
+        case "":
+            return new EmptyCommand();
+
+        case "exit":
+        case "bye":
+            return new ExitCommand();
+
+        // show list
+        case "list":
```
Perhaps the unnecessary empty lines could be removed?

marquestye commented on src/main/java/duke/parse/command/Command.java:
```diff
@@ -0,0 +1,16 @@
+     * @return whether the execution allows the program to continue,
+     *     true if it can, false means the program must exit
+     */
+    boolean execute(Duke bot);
```
Missing a parameter description.

marquestye commented on src/main/java/duke/parse/command/EchoCommand.java:
```diff
@@ -0,0 +1,44 @@
+     * @return
+     */
+    @Override
+    public boolean equals(Object another) {
```
Incomplete javadocs.

marquestye commented on src/main/java/duke/parse/command/ListCommand.java:
```diff
@@ -0,0 +1,83 @@
+            return this.isExcludingDone == anotherList.isExcludingDone
+                    && ((this.date == null && anotherList.date == null)
+                        || this.date.equals(anotherList.date))
+                    && this.type.equals(anotherList.type);
```
Perhaps you could split up the boolean checking to reduce complicated expressions.

marquestye commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,141 @@
+                throw new FileIoException();
+            }
+        }
+    }
```
Consider breaking up this method, as per the guidelines on [avoiding long methods](https://nus-cs2103-ay2324s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods).

marquestye commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,69 @@
+     * @return data representation of the task to be stored in the disk
+     */
+    @Override
+    public String data() {
```
Perhaps you rename this method to be in the form of a verb.

marquestye commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,158 @@
+    /**
+     * Thrown if a function is trying to access a task index out of range.
+     */
+    public static class TaskIndexOutOfRange extends Exception {
```
Good exception naming!

marquestye commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,158 @@
+     * @param date the date to include deadlines before and events happening on,
+     *             null if to not filter by date
+     */
+    private static String getTasks(ArrayList<Task> taskList, boolean isExcludingDone, LocalDate date) {
```
Perhaps you could adjust the parameter name to differentiate from private field which has the same name (taskList).

marquestye commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,158 @@
+     */
+    public String getTasks(boolean isExcludingDone, LocalDate date) {
+        ArrayList<Task> taskList = (ArrayList<Task>) this.taskList.clone();
+        return TaskList.getTasks(taskList, isExcludingDone, date);
```
Same here, maybe something like "taskListClone" might be clearer for readers?

marquestye commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,69 @@
+ * Represents a deadline.
+ * A deadline contains an end time.
+ */
+public class Deadline extends Task {
```
Consider renaming the class to DeadlineTask for better clarity. Same goes for EventTask and TodoTask.

marquestye commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,127 @@
+        DEADLINE,
+        EVENT,
+        DEFAULT
+    }
```
The second time this enum has been defined (also in ListCommand). Perhaps you could extract it out to its own class, and import when necessary!

marquestye commented on src/main/java/duke/ui/text/TextUi.java:
```diff
@@ -0,0 +1,256 @@
+        System.out.println(
+                "Alright, here is your list of "
+                + typeString
+                + (isExcludingDone ? " not done" : "")
+                + (date != null
+                        ? (type == Type.DEADLINE
+                            ? " before "
+                            : type == Type.EVENT
+                            ? " happening on "
+                            : " for "
+                        ) + DateTimeManager.dateToDisplay(date)
+                        : ""
+                ) + ":"
```
Consider breaking up this chain of ternary operators, by resolving the logic outside of the print statement first.

marquestye reviewed (commented):
LGTM, just keep in mind the coding standard on JavaDocs!

AaronJT1 commented on src/test/java/duke/parse/ParserTest.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Test
+    public void emptyTest() {
+        try {
+            assertEquals(new EmptyCommand(), Parser.parse(""));
+        } catch (Parser.ParseError e) {
+            notifyError();
+        }
+    }
```
Perhaps would it be better if we renamed the test from emptyTest to the format "featureUnderTest_testScenario_expectedBehavior()"?

AaronJT1 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,69 @@
+ * A deadline contains an end time.
+ */
+public class Deadline extends Task {
+    private LocalDateTime deadline;
```
Perhaps would it be better to name this variable to another name as it is same as the Deadline class and readers might get confused?

AaronJT1 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,94 @@
+    /**
+     * Marks the task as having been done.
+     */
+    public void markAsDone() {
```
I like how use a verb to name your methods!! 😄

AaronJT1 commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -0,0 +1,46 @@
+     * @return the string representation of this task in disk
+     */
+    @Override
+    public String data() {
```
Perhaps a method name like getDataToBeStored() is better, since it is a verb?

AaronJT1 reviewed (commented):
LGTM. Just minor naming nits here and there, but good job!!

nknguyenhc committed:
Add and enable assertions

nknguyenhc committed:
Improve code quality

Currently, there are a number of parts of the code that can be improved.

Hence there are changes to the code to make it easier to read.

Changes include:
* Format Javadoc for parameters.
* Break long methods into a combination of shorter methods
in Storage and Parser classes.
* Declare task type in Task interface and
remove task type declaration elsewhere.

This commit also fixes a bug where the command to list a type of task
displays all tasks without type filtering.

nknguyenhc committed:
Merge pull request #1 from nknguyenhc/branch-A-Assertions

Add and enable assertions

nknguyenhc committed:
Merge branch 'master' into branch-A-CodeQuality

nknguyenhc committed:
Merge pull request #2 from nknguyenhc/branch-A-CodeQuality

Improve code quality

nknguyenhc committed:
Add indentation for list of a day

Task done for add on: view deadlines before and/or events happening on a specific date.

nknguyenhc committed:
Option to run GUI or TextUI

nknguyenhc committed:
Add update functionality

nknguyenhc committed:
Merge branch 'branch-C-Update'

nknguyenhc committed:
Add Github Actions

Fix style issues after running `gradlew check`.

nknguyenhc committed:
Change FXML version to 17.0.7

nknguyenhc committed:
Better GUI

Changes include:
* Allow resizing of window
* Highlight error message with red text

Changes already implemented before this commit:
* Padding, font and clear text box
* Profile images

nknguyenhc committed:
Add more error handling

Changes include:
* Date time parse error (i.e. correct datetime format but invalid datetime)
is clearly indicated.
* Notify user of leading spaces.
* Add runtime error handling previous missed out.
* Prevent adding duplicate tasks.

nknguyenhc committed:
Merge branch 'branch-A-BetterGui'

nknguyenhc committed:
Merge branch 'branch-A-MoreErrorHandling'

nknguyenhc committed:
Update test case

nknguyenhc committed:
Update to follow Javadocs convention

nknguyenhc committed:
Publish user guide

User guide completed.

Minor changes:
* Add icon to the application
* Add bot name to the application
* Update Javadocs to reflect available commands

nknguyenhc committed:
Update Readme

Small change: no more markdown block in user guide.

nknguyenhc committed:
Corrected errors in readme
